core/src/storage/lazy/lazy_chunk.rs

```diff
@@ -0,0 +1,384 @@
+
+    /// Returns a shared reference to the value of the entry.
+    pub fn value(&self) -> Option<&T> {
+        self.value.as_ref()
+    }
+
+    /// Returns an exclusive reference to the value of the entry.
+    ///
+    /// # Note
+    ///
+    /// This changes the `mutate` state of the entry if the entry was occupied
+    /// since the caller could potentially change the returned value.
+    pub fn value_mut(&mut self) -> Option<&mut T> {
+        self.mutated.set(self.value.is_some());
+        self.value.as_mut()
+    }
+
+    /// Takes the value from the entry and returns it.
+    ///
+    /// # Note
+    ///
+    /// This changes the `mutate` state of the entry if the entry was occupied.
+    pub fn take_value(&mut self) -> Option<T> {
+        self.mutated.set(self.value.is_some());
+        self.value.take()
+    }
+
+    /// Puts the new value into the entry and returns the old value.
+    ///
+    /// # Note
+    ///
+    /// This changes the `mutate` state of the entry to `true` as long as at
+    /// least one of `old_value` and `new_value` is `Some`.
+    pub fn put(&mut self, new_value: Option<T>) -> Option<T> {
+        match new_value {
+            Some(new_value) => {
+                self.mutated.set(true);
+                self.value.replace(new_value)
+            }
+            None => self.take_value(),
+        }
+    }
+}
+
+impl<T> LazyChunk<T> {
+    /// Creates a new empty lazy chunk that cannot be mutated.
+    pub fn new() -> Self {
+        Self {
+            key: None,
+            cached_entries: UnsafeCell::new(EntryMap::new()),
+        }
+    }
+
+    /// Returns a shared reference to the entries.
+    fn entries(&self) -> &EntryMap<T> {
+        // SAFETY: We return a shared reference while `entries` is a method
+        //         with a `&self` receiver so we don't break lifetime or borrow
+        //         rules in isolation.
+        unsafe { &*self.cached_entries.get() }
+    }
+}
+
+impl<T> StorageSize for LazyChunk<T>
+where
+    T: StorageSize,
+{
+    const SIZE: u64 = <T as StorageSize>::SIZE * (core::u32::MAX as u64);
+}
+
+impl<T> PullForward for LazyChunk<T>
+where
+    T: StorageSize,
+{
+    fn pull_forward(ptr: &mut KeyPtr) -> Self {
+        Self {
+            key: Some(ptr.next_for::<Self>()),
+            cached_entries: UnsafeCell::new(BTreeMap::new()),
+        }
+    }
+}
+
+impl<T> PushForward for LazyChunk<T>
+where
+    T: PushForward + StorageSize,
+{
+    fn push_forward(&self, ptr: &mut KeyPtr) {
+        let key = ptr.next_for::<Self>();
+        let elem_size = <T as StorageSize>::SIZE;
+        assert_eq!(self.key, Some(key));
+        for (&index, entry) in self.entries().iter().filter(|(_, entry)| entry.mutated())
+        {
+            let offset: Key = key + (index as u64 * elem_size);
+            let mut ptr = KeyPtr::from(offset);
+            PushForward::push_forward(&**entry, &mut ptr);
+        }
+    }
+}
+
+impl<T> LazyChunk<T>
+where
+    T: scale::Decode,
```

Comment:
`StorageSize + PullForward`?

Reply:
I am currently working that out. Not yet done.